File: src/main/java/net/artux/pdanetwork/service/communication/ConversationServiceImpl.java
package net.artux.pdanetwork.service.communication;

import lombok.RequiredArgsConstructor;
import net.artux.pdanetwork.entity.communication.ConversationEntity;
import net.artux.pdanetwork.entity.user.UserEntity;
import net.artux.pdanetwork.models.communication.CommunicationMapper;
import net.artux.pdanetwork.models.communication.ConversationCreateDTO;
import net.artux.pdanetwork.models.communication.ConversationDTO;
import net.artux.pdanetwork.models.page.QueryPage;
import net.artux.pdanetwork.models.page.ResponsePage;
import net.artux.pdanetwork.repository.comminication.ConversationRepository;
import net.artux.pdanetwork.repository.user.UserRepository;
import net.artux.pdanetwork.service.user.UserService;
import net.artux.pdanetwork.service.util.PageService;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.LinkedList;
import java.util.List;
import java.util.Set;
import java.util.UUID;

@Service
@Transactional
@RequiredArgsConstructor
public class ConversationServiceImpl implements ConversationService {

    private final CommunicationMapper mapper;
    private final ConversationRepository repository;
    private final UserService userService;
    private final UserRepository userRepository;
    private final PageService pageService;

    @Override
    public ConversationDTO createConversation(ConversationCreateDTO createDTO) {
        ConversationEntity conversationEntity = new ConversationEntity();

        if(createDTO.getType() == ConversationEntity.Type.PRIVATE && createDTO.getUsers().size() != 1) {
            throw new RuntimeException("Необходимо указать id собеседника");
        }

        conversationEntity.setMembers(userRepository.findAllByIdIn(createDTO.getUsers()));
        conversationEntity.getMembers().add(userService.getUserById());

        conversationEntity.setTitle(createDTO.getTitle());
        conversationEntity.setIcon(createDTO.getIcon());
        conversationEntity.setOwner(userService.getUserById());
        conversationEntity.setType(createDTO.getType());
        conversationEntity.setTime(Instant.now());

        return mapper.dto(repository.save(conversationEntity));
    }

    @Override
    public ConversationDTO editConversation(UUID id, ConversationCreateDTO createDTO) {
        ConversationEntity conversationEntity = repository.findByIdAndOwner(id, userService.getUserById()).orElseThrow();

        if(createDTO.getType() == ConversationEntity.Type.PRIVATE && createDTO.getUsers().size() != 1) {
            throw new RuntimeException("Необходимо указать id собеседника");
        }

        conversationEntity.setMembers(userRepository.findAllByIdIn(createDTO.getUsers()));
        conversationEntity.getMembers().add(userService.getUserById());

        conversationEntity.setTitle(createDTO.getTitle());
        conversationEntity.setIcon(createDTO.getIcon());
        conversationEntity.setType(createDTO.getType());
        conversationEntity.setTime(Instant.now());

        return mapper.dto(repository.save(conversationEntity));
    }

    @Override
    public ConversationDTO getConversation(UUID id) {
        return repository.findByIdAndMembersContains(id, userService.getUserById()).map(mapper::dto).orElseThrow();
    }

    @Override
    public Page<ConversationDTO> getConversations(Pageable queryPage) {
        UserEntity user = userService.getUserById();
        Page<ConversationEntity> conversationEntityPage = repository.findAllByMembersContains(user, queryPage);

        conversationEntityPage.map(conversation -> {
            if(conversation.getType() == ConversationEntity.Type.PRIVATE) {
                UserEntity companion = getUniqueUser(userService.getCurrentId(), conversation.getMembers());
                conversation.setTitle(companion.getName() + " " + companion.getNickname());
                conversation.setIcon(companion.getAvatar());
            }
            return conversation;
        });

        return conversationEntityPage.map(mapper::dto);
    }

    @Override
    public ResponsePage<ConversationDTO> getConversations(QueryPage queryPage) {
        Page<ConversationDTO> conversationEntityPage = getConversations(pageService.getPageable(queryPage));
        return pageService.mapDataPageToResponsePage(conversationEntityPage, conversationEntityPage.getContent());
    }

    @Override
    public boolean deleteConversation(UUID id) {
        ConversationEntity entity = repository.findById(id).orElseThrow();
        UserEntity user = userService.getUserById();
        if (entity.getType().equals(ConversationEntity.Type.PRIVATE) && entity.getMembers().contains(user)) {
            repository.delete(entity);
            return true;
        } else if (entity.getOwner().equals(user)) {
            repository.delete(entity);
            return true;
        }
        return false;
    }

    @Override
    public ConversationEntity getPrivateConversation(UUID pda1, UUID pda2) {
        //Optional<ConversationEntity> conversation = repository.findByMembersContainsAndType(ConversationEntity.Type.PRIVATE, pda1, pda2);
        return null; // TODO
    }

    @Override
    public ConversationDTO getConversationEntity(UUID id) {
        return mapper.dto(repository.findById(id).orElse(null));
    }

    public ConversationEntity getConversationByIdForUser(UUID id, UserEntity user) {
        return repository.findByIdAndMembersContains(id, user).orElseThrow();
    }

    @Override
    public ConversationEntity createPrivateConversation(UUID pdaId, UUID id) {
        ConversationEntity conversationEntity = new ConversationEntity();
        List<UUID> list = new LinkedList<>();
        list.add(pdaId);
        list.add(id);
        conversationEntity.setMembers(userRepository.findAllByIdIn(list));
        conversationEntity.setTitle("");
        conversationEntity.setIcon("");
        conversationEntity.setOwner(userService.getUserById());
        conversationEntity.setType(ConversationEntity.Type.PRIVATE);
        conversationEntity.setTime(Instant.now());
        return repository.save(conversationEntity);
    }


    private UserEntity getUniqueUser(UUID userId, Set<UserEntity> users) {
        return users.stream().filter(user -> !user.getId().equals(userId)).findFirst().orElse(null);
    }
}

File: pdanet/src/main/java/net/artux/pdanetwork/service/communication/ConversationServiceImpl.kt
package net.artux.pdanetwork.service.communication

import net.artux.pdanetwork.dto.page.QueryPage
import net.artux.pdanetwork.dto.page.ResponsePage
import net.artux.pdanetwork.entity.communication.ConversationEntity
import net.artux.pdanetwork.entity.mappers.CommunicationMapper
import net.artux.pdanetwork.entity.user.UserEntity
import net.artux.pdanetwork.exception.UserNotFoundException
import net.artux.pdanetwork.models.communication.ConversationCreateDTO
import net.artux.pdanetwork.models.communication.ConversationDTO
import net.artux.pdanetwork.models.communication.ConversationType
import net.artux.pdanetwork.repository.comminication.ConversationRepository
import net.artux.pdanetwork.repository.comminication.MessageRepository
import net.artux.pdanetwork.repository.user.UserRepository
import net.artux.pdanetwork.service.user.UserService
import net.artux.pdanetwork.service.util.PageService
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID

@Service
@Transactional
@Suppress("TooManyFunctions")
open class ConversationServiceImpl(
    private val mapper: CommunicationMapper,
    private val repository: ConversationRepository,
    private val messageRepository: MessageRepository,
    private val userService: UserService,
    private val userRepository: UserRepository,
    private val pageService: PageService,
) : ConversationService {

    override fun createConversation(createDTO: ConversationCreateDTO): ConversationDTO {
        val conversationEntity = ConversationEntity()
        validateConversation(createDTO, conversationEntity)
        return mapper.dto(repository.save(conversationEntity))
    }

    override fun editConversation(id: UUID, createDTO: ConversationCreateDTO): ConversationDTO {
        val conversationEntity = repository.findByIdAndOwner(id, userService.getCurrentUser()).orElseThrow()
        validateConversation(createDTO, conversationEntity)
        return mapper.dto(repository.save(conversationEntity))
    }

    override fun getConversation(id: UUID): ConversationDTO {
        return repository.findByIdAndMembersContains(id, userService.getCurrentUser())
            .map { entity: ConversationEntity? -> mapper.dto(entity) }
            .orElseThrow()
    }

    override fun getConversationWithUser(userId: UUID): List<ConversationDTO> {
        // TODO сделать через sql запрос
        val conversations =
            repository.findAllByTypeAndMembers_Id(ConversationEntity.Type.PRIVATE, userService.getCurrentId())
        return conversations.stream().filter { conversation: ConversationEntity ->
            if (conversation.type == ConversationEntity.Type.PRIVATE) {
                return@filter conversation.members.contains(userService.getCurrentUser(userId))
            }
            false
        }.map { entity: ConversationEntity? -> mapper.dto(entity) }.toList()
    }

    override fun getConversations(queryPage: Pageable): Page<ConversationDTO> {
        val user = userService.getCurrentUser()
        val conversationEntityPage = repository.findAllByMembersContains(user, queryPage)
        conversationEntityPage.map { conversation: ConversationEntity ->
            if (conversation.type == ConversationEntity.Type.PRIVATE) {
                val companion = getUniqueUser(userService.getCurrentId(), conversation.members)
                conversation.title = companion.name + " " + companion.nickname
                conversation.icon = companion.avatar
            }
            conversation
        }
        return conversationEntityPage.map { entity: ConversationEntity? -> mapper.dto(entity) }
    }

    override fun getConversations(queryPage: QueryPage): ResponsePage<ConversationDTO> {
        val conversationEntityPage = getConversations(
            pageService.getPageable(queryPage)
        )
        return pageService.mapDataPageToResponsePage(conversationEntityPage, conversationEntityPage.content)
    }

    @Suppress("ReturnCount")
    override fun deleteConversation(id: UUID): Boolean {
        val entity = repository.findById(id).orElseThrow()
        val user = userService.getCurrentUser()
        if (entity.type == ConversationEntity.Type.PRIVATE && entity.members.contains(user)) {
            messageRepository.deleteAllByConversation(entity)
            repository.delete(entity)
            return true
        } else if (entity.owner == user) {
            messageRepository.deleteAllByConversation(entity)
            repository.delete(entity)
            return true
        }
        return false
    }

    override fun getConversationEntity(id: UUID): ConversationDTO {
        return mapper.dto(repository.findById(id).orElse(null))
    }

    override fun deleteAllConversationsWithUser(id: UUID): Boolean {
        val user = userService.getCurrentUser()
        messageRepository.deleteAllByAuthor(user)
        repository.deleteAllByMembersContainsAndTypeEquals(user, ConversationEntity.Type.PRIVATE)
        return true
    }

    private fun getUniqueUser(userId: UUID, users: Set<UserEntity>): UserEntity {
        return users.stream().filter { user: UserEntity -> user.id != userId }.findFirst().orElseThrow {
            UserNotFoundException("Пользователь не найден")
        }
    }

    private fun validateConversation(createDTO: ConversationCreateDTO, conversationEntity: ConversationEntity) {
        if (createDTO.type == ConversationType.PRIVATE && createDTO.users.size != 1) {
            throw UserNotFoundException("Необходимо указать id собеседника")
        }
        conversationEntity.members = userRepository.findAllByIdIn(createDTO.users)
        conversationEntity.members.add(userService.getCurrentUser())
        conversationEntity.title = createDTO.title
        conversationEntity.icon = createDTO.icon
        conversationEntity.owner = userService.getCurrentUser()
        conversationEntity.type = ConversationEntity.Type.valueOf(createDTO.type.name)
        conversationEntity.time = Instant.now()
    }
}
